src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -164,127 +168,161 @@ private static double round(double value) {
+     * | () | _ / __|| . |\ V / _ \| |_| |_| | ._|
+     * \___/|_| |___|_|\_| \_/_/ \_\___|\___/|___|
+     * ---------------- BETTER SOFTWARE, FASTER --
+     *
+     * https://www.openvalue.eu/
+     *
+     * Made you look.
+     *
+     */
+
+    private static final long DOT_BITS = 0x10101000;
+    private static final long MAGIC_MULTIPLIER = (100 * 0x1000000 + 10 * 0x10000 + 1);
+
+    private static int extractTemp(final int decimalSepPos, final long invNumberBits, final long numberBits) {
+        // Awesome idea of merykitty:
+        int min28 = (28 - decimalSepPos);
+        // Calculates the sign
         final long signed = (invNumberBits << 59) >> 63;
         final long minusFilter = ~(signed & 0xFF);
-        final long digits = ((numberBits & minusFilter) << (28 - decimalSepPos)) & 0x0F000F0F00L;
-        final long absValue = ((digits * MAGIC_MULTIPLIER) >>> 32) & 0x3FF; // filter just the result
+        // Use the pre-calculated decimal position to adjust the values
+        final long digits = ((numberBits & minusFilter) << min28) & 0x0F000F0F00L;
+        // Multiply by a magic (100 * 0x1000000 + 10 * 0x10000 + 1), to get the result
+        final long absValue = ((digits * MAGIC_MULTIPLIER) >>> 32) & 0x3FF;
+        // And perform abs()
         final int temp = (int) ((absValue + signed) ^ signed); // non-patented method of doing the same trick
         return temp;
     }
 
+    private static final long SEPARATOR_PATTERN = 0x3B3B3B3B3B3B3B3BL;
+
+    // Takes a long and finds the bytes where this exact pattern is present.
+    // Cool bit manipulation technique: SWAR (SIMD as a Register).
     private static long getDelimiterMask(final long word) {
-        long match = word ^ SEPARATOR_PATTERN;
-        return (match - 0x0101010101010101L) & ~match & 0x8080808080808080L;
+        final long match = word ^ SEPARATOR_PATTERN;
+        return (match - 0x0101010101010101L) & (~match & 0x8080808080808080L);
+        // I've put some brackets separating the first and second part, this is faster.
+        // Now they run simultaneous after 'match' is altered, instead of waiting on each other.
     }
 
-    private static final long SEPARATOR_PATTERN = 0x3B3B3B3B3B3B3B3BL;
-    private static final long DOT_BITS = 0x10101000;
-    private static final long MAGIC_MULTIPLIER = (100 * 0x1000000 + 10 * 0x10000 + 1);
-
     /**
      * For case multiple hashes are equal (however unlikely) check the actual key (using longs)
      */
-    static boolean arrayEquals(final long[] a, final long[] b, final int length) {
-        for (int i = 0; i < length; i++) {
-            if (a[i] != b[i])
+    private static boolean memoryEqualsEntry(final long startAddress, final long[] entry, final long finalBytes, final int amountLong) {
+        for (int i = 0; i < (amountLong - 1); i++) {
+            int step = i << 3; // step by 8 bytes
+            if (UNSAFE.getLong(startAddress + step) != entry[i])
```

Comment:
I think the compiler might not be able to proof that (i < entry.length) here at all times and so maybe also unsafe is a minor improvement for the array access here. Once you go unsafe, you have to be all in ;-).

I kind of liked that buffer array before btw, and was surprised by how fast this worked.

Reply:
I've already done just that; I've been keeping two branches locally with much more Unsafe; and until just now they were slower on my machine. However, with the latest changes, it is now actually faster, so I'm switching over to more and more Unsafe method calls and doing only memory.

I do still copy the bytes into the table, I think this will improve speed for the larger dataset specifically.